drivers: kscan: generalize and improve FT5336 driver

gmarull opened the pull request:
The FT5336 driver has been improved and generalized so that multiple
Focaltech touch controllers can be supported using the same single
driver. According to specifications the following list of controller
variants should be supported: FT5x06, FT5606, FT5x16, FT6x06, Ft6x36,
FT5x06i, FT5336, FT3316, FT5436i, FT5336i and FT5x46.

Tested using ER-TFTM028-4 display module (FT6X06).

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
drivers: kscan: generalize and improve FT5336 driver

The FT5336 driver has been improved and generalized so that multiple
Focaltech touch controllers can be supported using the same single
driver. According to specifications the following list of controller
variants should be supported: FT5x06, FT5606, FT5x16, FT6x06, Ft6x36,
FT5x06i, FT5336, FT3316, FT5436i, FT5336i and FT5x46.

Tested using ER-TFTM028-4 display module (FT6X06).

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

jfischer-no committed:
shields: adafruit_2_8_tft_touch_v2: enable touch input support

Enable FT6x06 controller driver and touch input support.

Signed-off-by: Johann Fischer <j.fischer@phytec.de>

jfischer-phytec-iot commented:
Looks good at first glance. Actually CI does not build this driver. I tested it on adafruit_2_8_tft_touch_v2 and it works (there are few error log regarding i2c_nrfx_twi and input data queue sometimes). I would like you to cherry-pick 6da1d4d03b452e4b8b1834fe0351d76d0ec02860 from https://github.com/jfischer-phytec-iot/zephyr/tree/pr-ft6x06_adafruit_2_8_tft_touch_v2 (with the CI would build the driver for nrf52840dk_nrf52840).

gmarull commented:
Thanks @jfischer-phytec-iot, will pick it. I think it would be worth adding to the documentation that interrupt mode can be used if this jumper is soldered (similar thing happens for ER-TFTM028-4 at least):

![image](https://user-images.githubusercontent.com/25011557/88299467-ad56d880-cd02-11ea-9abf-dc0e945f98a5.png)

When in polling mode, maybe the polling period can be increased so that no message queues are lost (I suppose they are consumed at < 10ms).

gmarull committed:
shields: adafruit_2_8_tft_touch_v2: add note about touch IRQ

Add note about the availability of touch panel IRQ linei when a jumper
in the board is soldered. A Device Tree commented entry has also been
added to ease testing it.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

MaureenHelm reviewed (commented):
Can this be combined with the ft5336 driver? There is a lot of code in common.

gmarull commented:
@MaureenHelm Actually looking at ft5336 this driver should be compatible as is, the features we access use the same registers, field positions... Could you check?

MaureenHelm commented:
> @MaureenHelm Actually looking at ft5336 datasheet this driver should be compatible as is, the features we access use the same registers, field positions... Could you check?
> 
> In fact, looking at [focaltech-systems/drivers-input-touchscreen-FTS_driver](https://github.com/focaltech-systems/drivers-input-touchscreen-FTS_driver) linux code, FT5x06，FT5606，FT5x16，FT6x06，Ft6x36，FT5x06i，FT5336，FT3316，FT5436i，FT5336i，FT5x46 should be covered... Any suggestions on how to reflect that on a single driver...?

Take a look at how the fxos8700 sensor driver also supports mma8451, mma8652, and mma8653 sensors. We don't yet support multiple compatibles in the dts bindings, so we currently use a list of compatibles for boards that have the mma8xxx variant.

https://github.com/zephyrproject-rtos/zephyr/blob/41da0769c19174dbf603e2f9e027b126a446a9b3/boards/arm/reel_board/dts/reel_board.dtsi#L123-L129

I suggest we follow the fxos8700 example. I'd prefer not to rename the ft5336 driver to something more generic as that will break out-of-tree application configurations.

gmarull commented:
@MaureenHelm This is a device tree entry that lists multiple compatibles, but still `nxp,fxos8700` driver supports a single compatible. Afaik there is this driver that supports multiple compatibles: https://github.com/zephyrproject-rtos/zephyr/blob/master/drivers/eeprom/eeprom_at2x.c

gmarull commented:
@MaureenHelm @jfischer-phytec-iot I have just improved FT5336 driver making explicit that it supports multiple controllers.

erwango reviewed (approved):
Approved on shields side

andrewboie commented on boards/shields/adafruit_2_8_tft_touch_v2/adafruit_2_8_tft_touch_v2.overlay:
```diff
@@ -17,3 +17,13 @@
+		compatible = "focaltech,ft5336";
+		reg = <0x38>;
+		label = "FT5336";
+		/* Uncomment if IRQ line is available (requires to solder jumper) */
```
wondering if there's any way to conditionally do this in DTS via Kconfig, rather than requiring source tree modification? @galak

jfischer-phytec-iot replied:
This can be solved in the driver code, KSCAN_FT5336_INTERRUPT could be removed and DT_NODE_EXISTS(...) used instead.

jfischer-phytec-iot replied:
@gmarull Please consider to remove KSCAN_FT5336_INTERRUPT and use DT_NODE_EXISTS(...) here or in a subsequent PR.

gmarull replied:
@jfischer-phytec-iot Thanks for the suggestion. I'll try to improve using this macro in another PR as I also want to add the overlay for the test board I'm using.

tyrantlizard77 reviewed (commented):
looks fine only thing that stood out was the commented line in the DTS file, maybe there's a better way

carlescufi commented:
@gmarull please rebase

gmarull commented:
@carlescufi done
